common/sgx/eeid.c

```diff
@@ -0,0 +1,451 @@
+    memset(&blank_pg, 0, sizeof(blank_pg));
+    memset(&stack_pg, 0xcc, sizeof(stack_pg));
+
+#define ADD_PAGE(PG, T)                                      \
+    {                                                        \
+        OE_CHECK(oe_sgx_measure_load_enclave_data(           \
+            &hctx,                                           \
+            base,                                            \
+            base + vaddr,                                    \
+            (uint64_t)&PG,                                   \
+            SGX_SECINFO_REG | SGX_SECINFO_R | SGX_SECINFO_W, \
+            T));                                             \
+        vaddr += OE_PAGE_SIZE;                               \
+    }
+
+    uint64_t vaddr = eeid->vaddr;
+
+    if (with_eeid_pages)
+    {
+        oe_page_t fst_page;
+        *((uint64_t*)fst_page.data) =
+            0xEE1DEE1DEE1DEE1D; // A non-eeid enclave would segfault (no heap)
+                                // or see a 0.
+        size_t num_bytes = sizeof(oe_eeid_t) + eeid->data_size;
+        size_t num_pages =
+            num_bytes / OE_PAGE_SIZE + (num_bytes % OE_PAGE_SIZE) ? 1 : 0;
+        memcpy(fst_page.data + sizeof(uint64_t), eeid, num_bytes);
+        for (size_t i = 0; i < num_pages; i++)
+            ADD_PAGE(fst_page + i * OE_PAGE_SIZE, false);
+    }
+
+    for (size_t i = 0; i < eeid->size_settings.num_heap_pages; i++)
+        ADD_PAGE(blank_pg, false);
+
+    for (size_t i = 0; i < eeid->size_settings.num_tcs; i++)
+    {
+        vaddr += OE_PAGE_SIZE; /* guard page */
+
+        for (size_t i = 0; i < eeid->size_settings.num_stack_pages; i++)
+            ADD_PAGE(stack_pg, true);
+
+        vaddr += OE_PAGE_SIZE; /* guard page */
+
+        sgx_tcs_t* tcs;
+        memset(&tcs_pg, 0, sizeof(tcs_pg));
+        tcs = (sgx_tcs_t*)&tcs_pg;
+        tcs->flags = 0;
+        tcs->ossa = vaddr + OE_PAGE_SIZE;
+        tcs->cssa = 0;
+        tcs->nssa = 2;
+        tcs->oentry = eeid->entry_point;
+        tcs->fsbase = vaddr + (5 * OE_PAGE_SIZE);
+        tcs->gsbase = tcs->fsbase;
+        tcs->fslimit = 0xFFFFFFFF;
+        tcs->gslimit = 0xFFFFFFFF;
+
+        OE_CHECK(oe_sgx_measure_load_enclave_data(
+            &hctx,
+            base,
+            base + vaddr,
+            (uint64_t)&tcs_pg,
+            SGX_SECINFO_TCS,
+            true));
+
+        vaddr += OE_PAGE_SIZE;
+
+        for (size_t i = 0; i < 2; i++)
+            ADD_PAGE(blank_pg, true);
+        vaddr += OE_PAGE_SIZE; // guard
+        for (size_t i = 0; i < 2; i++)
+            ADD_PAGE(blank_pg, true);
+    }
+
+    oe_sha256_final(&hctx, cpt_mrenclave);
+
+done:
+    return OE_OK;
+}
+
+static bool is_zero(const uint8_t* buf, size_t sz)
+{
+    while (sz != 0)
+        if (buf[--sz] != 0)
+            return false;
+    return true;
+}
+
+oe_result_t verify_eeid(
+    const uint8_t* r_mrenclave,
+    const uint8_t* r_mrsigner,
+    uint16_t r_product_id,
```

Comment:
Please spell out the names, such as `reported_product_id`.

Reply:
Done.